arroyo/dlq.py

```diff
@@ -192,3 +232,51 @@ def reset(self) -> None:
         Reset the buffer.
         """
         self.__buffered_messages = defaultdict(deque)
+
+
+class InvalidMessageState:
+    """
+    This class is designed to be used internally by processing strategies to
+    store invalid messages pending commit.
+
+    If strict_ordering is True, an exception is raised if any invalid messages
+    are not in order.
+    """
+
+    def __init__(self, strict_ordering: bool = True) -> None:
+        self.__strict_ordering = strict_ordering
+        self.__invalid_messages: MutableSequence[InvalidMessage] = []
+
+    def __len__(self) -> int:
+        return len(self.__invalid_messages)
+
+    def append(self, invalid_message: InvalidMessage) -> None:
+        """
+        Mark the invalid message as committed so other strategies in the pipeline
+        don't try to commit the same offset when the exception is reraised.
+        """
+        invalid_message.needs_commit = False
+        self.__invalid_messages.append(invalid_message)
+
+    def build(self) -> Optional[Message[FilteredPayload]]:
+        """
+        Returns a filtered message to be committed down the line. If there is
+        nothing to commit, return None.
+        """
+        committable: MutableMapping[Partition, int] = {}
+        for m in self.__invalid_messages:
+            next_offset = m.offset + 1
+            if self.__strict_ordering and m.partition in committable:
+                if next_offset < committable[m.partition]:
+                    raise InvalidMessageOutOfOrder
```

Comment:
Any reason for this to be checked when building the filtered payload instead of when an invalid message is appended to the state ?

Reply:
No I don't believe so, it might be better to raise from `append()` instead